tutorials/2023-07-05-neo3-boa-writing-a-fungible-token-contract-using-python/index.md

```diff
@@ -0,0 +1,341 @@
+
+    if tokens_sender < amount:
+        return False
+
+    # if `amount` is 0 or the sender is also the receiver, there is no need to change the storage
+    if amount != 0 and from_address != to_address:
+        # use the `storage.put` method to change the storage
+        storage.put(from_address, type_helper.to_bytes(tokens_sender - amount))
+        storage.put(to_address, type_helper.to_bytes(tokens_receiver + amount))
+    
+    # if the method succeeds, it must fire the Transfer event (more details in the next section)
+    Nep17TransferEvent(from_address, to_address, amount)
+    # if the to_address is a smart contract, it must call the onNEP17Payment
+    if blockchain.get_contract(to_address) is not None:
+        contract.call_contract(to_address, 'onNEP17Payment', [from_address, amount, data])
+
+    return True
+```
+
+## 8. NEP-17 Event
+
+The NEP-17 standard defines a single event called `Transfer` that must be triggered when a transfer occurs.
+You can use the `Nep17TransferEvent` from Neo3-boa or create your own event with `CreateNewEvent`.
+
+In this example, we use this event twice: when minting all the tokens at the deploy and whenever a transfer operation occurs.
+
+```python
+# update coin.py adding the following code:
+from boa3.builtin.compile_time import CreateNewEvent
+
+on_nep17_transfer = CreateNewEvent(
+    [
+        ('from', Union[UInt160, None]),
+        ('to', Union[UInt160, None]),
+        ('amount', int),
+    ],
+    'Transfer'
+)
+
+# to trigger the event, just call it by passing the parameters
+# using the already implemented Event:
+Nep17TransferEvent(from_address, to_address, amount)
+# using the newly created Event:
+on_nep17_transfer(from_address, to_address, amount)
+```
+
+## 9. NEP-17 Callbacks
+
+The NEP-17 standard defines a single callback called `onNEP17Payment` that must be called if the recipient is a contract.
+This callback is used to notify the recipient that it has received tokens. It's up to the recipient to implement this
+callback. The recipient can reject the transfer by raising an exception.
+
+```python
+# update coin.py adding the following code:
+from boa3.builtin.contract import abort
+
+@public(name='onNEP17Payment')
+def on_nep17_payment(from_address: UInt160, amount: int, data: Any):
+    abort()     # in this example, the smart contract is rejecting all transfers made to it
+```
+
+## 10. Manifest Metadata
+
+Neo3-boa allows you to define the contract metadata using the `metadata` decorator. This information is used to generate
+the contract manifest file.
+
+In our example, we will define the contract name and the supported standards, but you can add any information you want
+to. We are also adding permissions to allow our contract to call the `onNEP17Payment` method. The compiler automatically
+does this step, but it's a good practice to define the permissions in the contract code.
+
+> You should add this method at the beginning of your file, before any other method.
```

Comment:
É obrigatório ficar em cima do arquivo ou é só convenção? Se for convenção é bom falar explicitamente q é convenção, se não for convenção acho q precisamos deixar mais forte essa mensagem (e cogitar mudar no neoboa)